core/storage/user/gae_models.py

```diff
@@ -97,6 +97,61 @@ def get_by_normalized_username(cls, normalized_username):
             cls.normalized_username == normalized_username).get()
 
 
+class ActivitiesCompletedByLearnerModel(base_models.BaseModel):
+    """Keeps track of all the explorations and collections completed by the
+    learner.
+
+    Instances of this class are keyed by the user id.
+    """
+    # IDs of all the explorations completed by the user.
+    completed_exp_ids = ndb.StringProperty(repeated=True, indexed=True)
+    # IDs of all the collections completed by the user.
+    completed_collection_ids = ndb.StringProperty(repeated=True, indexed=True)
+
+
+class IncompleteCollectionsModel(base_models.BaseModel):
+    """Keeps track of all the collections currently being completed by the
+    learner.
+
+    Instances of this class are keyed by the user id.
+    """
+    incomplete_collection_ids = ndb.StringProperty(
+        repeated=True, indexed=True)
+
+
+class IncompleteExplorationsModel(base_models.BaseModel):
+    """Keeps track of all the explorations currently being completed by the
+    learner.
+
+    Instances of this class are keyed by the user id.
+    """
+    # A list which stores the explorations partially completed by the learner.
+    # Each item in the list is a json object keyed by exploration id and the
+    # value corresponds to a few extra details -
+    # timestamp: When did the learner leave the exploration.
+    # version: Version of the exploration.
+    # state_name: The name of the state at which the learner left the
+    # exploration. Example -
+    # [
+    #  {
+    #   (exp_id_1): {
+    #    'timestamp': timestamp,
```

Comment:
Since you are storing blobs, I am concerned about possible schema changes in the future, e.g. what if we needed to store their history of answers submitted as well so that we can reproduce their previous session.

So, have you considered an alternative approach using ExpUserData model, where we store the timestamp, version and state name but using actual ndb properties? The exp_id and user_id would be indexed so we can do a get_by_user_id() call to fetch all such models corresponding to a given user_id using the precomputed indexes.

Or, yet another alternative: keep this restricted to just a list of ids, and in ExpUserData maintain "last playthrough" information separately. That way you at least have parity between all these models that keep track of ids. Not sure if there are other concerns here, though.